### DISCUSSION TITLE
점프를 두번이상 할 수 있는 방법

### DISCUSSION THREAD

**Discussion by JeongYire** [naverz/zepeto-studio-kor#307] [5. Q&A]
점프를 두번 이상 할 수 있게 만들어보려하는데,
ZEPETOCharacter 에서 Jump 메소드를 쓰게해도 땅에 닿을때 까지는 이 메소드가 실행이 되지 않습니다.
그렇다고 캐릭터컨트롤러 자체를 꺼버리면 아예 애니메이션이 정지되어버리니 끌 수도 없구요.

어느쪽 부터 건드려야 점프를 두번이상 할 수 있게 만들 수 있을까요?

**Comment by runasy-koonta** [comment]
점프 후 Gesture를 Idle상태로 만들고 다시 점프하면 안 되나요?

**Reply by JeongYire** [reply]
일단 ZepetoCharacter.ChangeStateAnimation(CharacterState.Idle); 를 하고 점프를 시도해도 땅에 떨어질때까지 점프가 불가능한건 여전합니다.
SetGesture를 하면 아예 애니메이션 전부가 멈춰버리고.. CancelGesture 는 안먹히네요

**Reply by runasy-koonta** [reply]
Reflection을 이용해 m_stateMachine에 접근해 강제로 컨트롤하면 가능하긴 한데, (기존의 State를 강제로 Idle로 바꿔준 뒤, 다시 새로운 State 설정해주면 됩니다.)
이 방법은 상당히 꼼수(?)같은 방법이라, 아무래도 @ x4cf 님의 방법이 가장 좋을 것 같아요.

또는, Animation을 재생하고 수동으로 점프해주는 방법도 있을 것 같습니다.

현재 ZEPETO의 점프 알고리즘은,
(0) 점프 액션 시작
(1) CurrentJumpPower = JumpPower (기본 Jump power + Additional jump power)
(2) CurrentJumpPower * Time.deltaTime 만큼 위로 이동
(3) CurrentJumpPower -= Gravity
(4) 바닥에 닿았는지 검사하고, 아니라면 (2)로 돌아감 (실제로는 Update문 안에서 이 작업이 수행될테니, 반복문을 사용하진 않겠네요)
(5) 점프 액션 끝
인 것으로 보여서, 구현 자체는 크게 어려워 보이지 않습니다.

**Reply by JeongYire** [reply]
방법을 알려주셔서 감사합니다. 멀티플레이 했을시 다른유저의 동작구현까지 보여줘야할거 생각하니 머리가 아파오네요...

**Comment by x4cf** [comment]
점프도중 점프버튼을 다시 눌렀을때 캐릭터 위치에 투명한 발판을 만들어서 착지시킨뒤 빠르게 다시 점프시키는 방법은 어떨까요?

**Reply by pda017** [reply]
ㅋㅋ 그렇게 구현해야 된다는게 너무 비참한데요. 제페토측에서 공중점프 기능을 만들어줬으면 좋겠네요.

**Reply by JeongYire** [reply]
그런 방법을 저도 생각하긴했는데, 아무래도 좀 불안하네요. 점프할때마다 계속 생성과 제거작업을 해야하기도하고...
